packages/require/lib/index.ts

```diff
@@ -0,0 +1,130 @@
+
+  expectOptions(options, ["file"]);
+
+  const conf = TruffleConfig.default().with(options);
+
+  const scriptModule = new Module(scriptPath);
+
+  // Provide all the globals listed here: https://nodejs.org/api/globals.html
+  const sandbox: { [index: string]: any } = {
+    __filename: path.basename(scriptPath),
+    __dirname: path.dirname(scriptPath),
+    Buffer,
+    clearImmediate,
+    clearInterval,
+    clearTimeout,
+    console,
+    exports: scriptModule.exports,
+    global,
+    process,
+    setImmediate,
+    setInterval,
+    setTimeout,
+    module: scriptModule,
+    config: conf,
+    artifacts: conf.resolver,
+    require: (id: string) => {
+      // Ugh. Simulate a full require function for the file.
+      id = id.trim();
+
+      // If absolute, just require.
+      if (path.isAbsolute(id)) return originalRequire(id);
+
+      // If relative, it's relative to the file.
+      if (id[0] === ".") {
+        return originalRequire(path.join(path.dirname(scriptPath), id));
+      } else {
+        // Not absolute, not relative, must be a globally or locally installed module.
+        // Try local first.
+        // Here we have to require from the node_modules directory directly.
+
+        var moduleDir = conf.working_directory;
+        while (true) {
+          try {
+            return originalRequire(path.join(moduleDir, "node_modules", id));
+          } catch (e) {}
+          var oldModuleDir = moduleDir;
+          moduleDir = path.dirname(moduleDir);
+          if (moduleDir === oldModuleDir) break;
+        }
+
+        // Try global, and let the error throw.
+        return originalRequire(id);
+      }
+    }
+  };
+
+  // Now add contract names.
+  Object.keys(conf.context || {}).forEach(key => {
+    sandbox[key] = conf.context[key];
+  });
+
+  const context = vm.createContext(sandbox);
+
+  const old_cwd = process.cwd();
+  const cwd = path.dirname(scriptPath);
+  process.chdir(cwd);
+
+  registerTsNode(conf, scriptPath, context);
+
+  const source = compile(conf, scriptPath);
+
+  const script = new vm.Script(source, { filename: scriptPath });
+
+  script.runInContext(context);
+  scriptModule.loaded = true;
+
+  const returnValue = scriptModule.exports.default ?? scriptModule.exports;
+
+  process.chdir(old_cwd);
+
+  return returnValue;
+}
+
+export function exec(options: ExecOptions, done: (...args: any[]) => void) {
+  expectOptions(options, [
+    "contracts_build_directory",
+    "file",
+    "resolver",
+    "provider",
+    "network",
+    "networks",
+    "network_id"
+  ]);
+
+  const interfaceAdapter = createInterfaceAdapter({
+    provider: options.provider,
+    networkType: options.networks[options.network].type
+  });
+  const web3 = new Web3Shim({
+    provider: options.provider,
+    networkType: options.networks[options.network].type
+  });
+
+  try {
+    const fn = this.file({
+      file: options.file,
+      context: { web3, interfaceAdapter },
```

Comment:
I just realized that while this codepath will attempt to compile TypeScript `truffle exec` scripts, it won't include the necessary type definitions for the `web3` and `interfaceAdapter` globals here.

I think that's fine personally, as we aren't claiming support for TypeScript in `truffle exec` here.

Reply:
I think TS support for `truffle exec` just needs its own feature ticket, yeah

Reply:
#5083